### PR TITLE
Use Dropbox's zxcvbn package to calculate password strength

### LINKED ISSUE #12910: Password strength estimation is completely bad
### What Grafana version are you using?
5.2.2

### What did you do?
I wanted to create an account after I received an email for that.

### What was the expected result?
Have an accurate information about password's strength.

### What happened instead?
It gave me completely wrong strength estimation, and is purely based on the length, considering tests and https://github.com/grafana/grafana/blob/master/public/app/core/components/PasswordStrength.tsx#L13-L30

No, "aaaa" does not have an average strength: it's bad. Same goes for "aaaaaaaaa", it is far from being a strong password.
Password strength isn't just a length question as you can see.

Libs probably exist that test strength, lists of known bad passwords and so on, but if it's just to estimate it given the length, imho it's better not show any information considering how wrong is it.
Making it longer as for https://github.com/grafana/grafana/issues/1892 by extending it from 6 to 8 isn't better.
No, "123456789" or "password1" aren't strong passwords and never will.

Please correct that or remove that estimation, it would be better than giving a wrong security feeling.

### PR DESCRIPTION
**Pull request by Lusitaniae**
Previous password strength was solely base on password length. 

It's now using Dropbox's library to calculate password strength.

More info: https://www.npmjs.com/package/zxcvbn

> More secure: policies often fail both ways, allowing weak passwords (P@ssword1) and disallowing strong passwords.

Fixes #12910.

### CODE REVIEW

**Comment by torkelo** [comment]
Thanks for contributing, we will try to review and merge this soon

**Comment by Limess** [comment]
It's worth noting that `zxcvbn.js` is a huge `400kB` gzipped. I'd highly recommend lazy loading this when used or loading it asynchronously on only the registration pages.

**Comment by Lusitaniae** [comment]
Thanks @Limess,

I've merged your PR to my branch

**Comment by Limess** [comment]
Hey @Lusitaniae - sorry about this! I believe changing the imports in 
to

```jsx
import React, {PureComponent} from 'react';
```

and removing the second line `zxcvbn` import which is now dynamic.

will fix up the tests and linting errors.

**Comment by Limess** [comment]
Looks like there's just something which isn't prettified correctly with [prettier](https://prettier.io/).

Running `npx prettier --write "**/*.{ts,tsx,scss}"` should solve that.

**Comment by Lusitaniae** [comment]
Would have seen it if I hadn't made the change from GitHub :laughing:

File: public/app/core/components/PasswordStrength.tsx [hunk 1]
```diff
@@ -1,36 +1,60 @@
...
 }
 
-export class PasswordStrength extends React.Component<Props, any> {
-  constructor(props) {
+export class PasswordStrength extends PureComponent<Props, any> {
+  constructor(props: Props) {
     super(props);
+
+    this.state = {
+      getScore(password: string) {
```
Comment (torkelo) [inline public/app/core/components/PasswordStrength.tsx]: This PR looks good except this, defining a function in the state object looks very unorthodox, have never seen it before. Also State is untyped so if you want to use state it should have a type. @dprokop have you seen this before? adding a function to the state?

Comment (Limess) [reply public/app/core/components/PasswordStrength.tsx]: I don't think this is particularly unorthodox - state, like props, is just a regular collection of whatever you want, functions not being special - the alternative I can see is to lazy load the whole component using zxcvbn and replace that, passing `getScore` as a prop, but that's just moving the logic upwards to the consumer.

Comment (dprokop) [reply public/app/core/components/PasswordStrength.tsx]: I haven't seen this approach being used neither. Not sure what the benefit of having this method in state is compared to regular instance method, bu this feels flaky. I see this being driven only by the dynamic import in `componentDidMount`. For that I suggest creating a higher order component that would serve as module loader and deliver the loaded module or part of its API to the desired component (PasswordStrength in this case)

**Comment by torkelo** [comment]
We have discussed this and not sure a password strength is needed for Grafana, most use external auth anyway and having this strength value is unnecessary. 

Improving this strength value means we should enforce it in the backend so we should really use a backend lib for this. And then we need to add a setting so admins can define their own level of strength etc. 

We decided to instead remove this strength value: https://github.com/grafana/grafana/issues/17748

**Pull request closed without merging**